activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -592,8 +592,10 @@ def affected_rows(raw_result)
         end
 
         def preprocess_query(sql)
-          check_if_write_query(sql)
-          mark_transaction_written_if_write(sql)
+          if write_query?(sql)
+            ensure_writes_are_allowed(sql)
+            mark_transaction_written
```

Comment:
Should we just inline both of these? They're both basically `x if y`, which was probably more method-worthy when they were called separately from each adapter etc? 🤷🏻‍♂️

Reply:
I suppose we could yes. More importantly we should pass down whether we expect a read or write query, e.g. if we're called from `select_all`, we should assume it's a read query.

That regexp is about 1.5% on the profile AFAICT.